### DISCUSSION TITLE
Best practice for importing node_modules when developing VS Code extention?

### DISCUSSION THREAD

**Discussion by Pixelherz23** [microsoft/vscode-discussions#928] [Extension Development QnA] [answered]
Hello! I am trying to display graphs in web view using 'maxGraph'.

I have tried different ways to import the lib.

Tried to import and use it in the extention.ts.Get:` Activating extension 'undefined_publisher.testextention' failed: navigator is not defined.`
I tried to split off the maxGraph logic in a separate typescript file and compile it in javascript => reference error as import/require() cause import/require() does not exits on client side.

Does anyone have some examples or can help me how to use libs in vs code extensions propely?

**Comment by r3m0t** [comment, marked as answer]
There are examples here

https://github.com/microsoft/vscode-webview-ui-toolkit-samples/tree/main

Basically, keep your webview code in separate files which can import libraries for browsers like maxGraph.

Have separate compilation for the webview code and the extension code.

**Reply by burkeholland** [reply]
@Pixelherz23 does this resolve your issue?

**Comment by olguzzar** [comment]
Going to mark this as answered since there's been a period of inactivity, but if that's not the case, please let us know!
